include/aws/common/linked_list.h

```diff
@@ -0,0 +1,86 @@
+#ifndef AWS_LINKED_LIST_H
+#define AWS_LINKED_LIST_H
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <stdint.h>
+
+#define aws_linked_list_of(type, name) struct name {                                                                   \
+    struct aws_linked_list_node list;                                                                                  \
+    type value;                                                                                                        \
+}                                                                                                                      \
+
+#define aws_container_of(ptr, type, member)  ((type *)ptr)->member
+#define aws_linked_list_node_of(ptr) (struct aws_linked_list_node *)ptr
+
+#define aws_linked_list_push_back(head, new_node) \
+    aws_linked_list_insert_before((head), (new_node))
+
+#define aws_linked_list_push_front(head, new_node) \
+    aws_linked_list_insert_after((head), (new_node))
+
+#define aws_linked_list_pop_back(head) \
+    aws_linked_list_remove((head)->prev)
+
+#define aws_linked_list_pop_front(head) \
+    head = aws_linked_list_remove(head);
+
+struct aws_linked_list_node {
+    struct aws_linked_list_node *next;
+    struct aws_linked_list_node *prev;
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+    * Initializes a doubly linked list.
+    */
+    AWS_COMMON_API void aws_linked_list_init(struct aws_linked_list_node *head);
+
+    /**
+    * Resets the list for either reuse or deletion.
```

Comment:
Resetting an intrusive list is not required for deletion.

It's better to just have init initialize the element to represent a list of a single element. Note also that `clear` and `init` have the same behavior, but the docs aren't clear at all about this, nor are they clear about the difference between the two.